internal/freelist/freelist.go

```diff
@@ -0,0 +1,109 @@
+package freelist
+
+import (
+	"sort"
+
+	"go.etcd.io/bbolt/internal/common"
+)
+
+type ReadWriter interface {
+	// Read calls Init with the page ids stored in te given page.
+	Read(page *common.Page)
+
+	// Write writes the freelist into the given page.
+	Write(page *common.Page)
+
+	// EstimatedWritePageSize returns the size of the freelist after serialization in Write.
+	// This should never underestimate the size.
+	EstimatedWritePageSize() int
+}
+
+type Interface interface {
+	ReadWriter
+
+	// Init initializes this freelist with the given list of pages.
+	Init(ids common.Pgids)
+
+	// Allocate returns the starting page id of a contiguous block of a given size in number of pages.
+	// If a contiguous block cannot be found then 0 is returned.
+	Allocate(txid common.Txid, numPages int) common.Pgid
+
+	// Count returns the number of free and pending pages.
+	Count() int
+
+	// FreeCount returns the number of free pages.
+	FreeCount() int
+
+	// FreePageIds returns all free pages.
+	FreePageIds() common.Pgids
+
+	// PendingCount returns the number of pending pages.
+	PendingCount() int
+
+	// AddReadonlyTXID adds a given read-only transaction id for pending page tracking.
+	AddReadonlyTXID(txid common.Txid)
+
+	// RemoveReadonlyTXID removes a given read-only transaction id from pending page tracking.
+	RemoveReadonlyTXID(txid common.Txid)
+
+	// FreePages releases any pages associated with closed read-only transactions.
+	FreePages()
```

Comment:
maybe to stay consistent with the other R-TX?

```suggestion
	FreeReadonlyTxPages()
```